### DISCUSSION TITLE
Next.js + WebSocket / Real-time examples

### DISCUSSION THREAD

**Discussion by bergkvist** [vercel/next.js#14950] [Help]
# Feature request

## I want to create a real-time application with Next.js

My use-case is that I want to download large files and perform computationally expensive work on them in the backend. To keep the client in sync with what is happening, I imagine using the WebSocket API is ideal (compared to a polling solution). This allows me to push progress updates/errors/results to the client as things happen.

For this I need a stateful backend (since this is required for websockets to work), with ability to use something like [NATS Streaming Server](https://github.com/nats-io/nats-streaming-server) or [REDIS Streams](https://redis.io/topics/streams-intro), [Apache Kafka](https://kafka.apache.org/) - or just [RxJs](https://github.com/ReactiveX/rxjs) within the server.

## Documentation on how to work with real-time data in Next.js seem to be non-existent

I'd love some examples of how this could be done in a good way. Especially with any nuances that might play into server-side vs client-side rendering together with streaming data.

**Comment by j-mcgregor** [comment]
I'm looking for stuff about this too but there's nothing - did you manage to find anything?

**Reply by leerob** [reply]
This might be helpful: https://www.roomservice.dev/

**Reply by PrincewillIroka** [reply]
@leerob your suggestion is nice, but room is an opinionated library. I'm looking for a more extensible option like integrating Socket.io with Next.js

**Reply by RobinYzewyn** [reply]
@PrincewillIroka did you find something with Socket.io and Nextjs? Or something else...

**Reply by PrincewillIroka** [reply]
No I didn't. On my own, I managed to integrate socket.io with Next.js on the client-side just the same way a normal React app would be integrated with socket.io.

**Reply by einSelbst** [reply]
twimc there once was this example available, but not anymore  https://github.com/vercel/next.js/tree/442fbfaa4d4e8205cf10c7e5d27b8f51b7a4bdc6/examples/with-socket.io

**Comment by gian412** [comment]
I'm doing some research on this too. As far as I understood, the only way to do something like this is by creating a custom server with, for example, express and integrating into it socket.io or similar. Then in the FE you can import socket.io-client and initialize it in the useEffect hook.
Any other idea anyone?

**Comment by gian412** [comment]
And just some minutes after my last comment, I think I found some resources that saved me (and maybe also to you) a lot of time.

At this [link](https://stackoverflow.com/questions/57512366/how-to-use-socket-io-with-next-js-api-routes), there is a solution that worked for me. As a disclaimer, I just found out and tried something really simple.

**Comment by dfjs** [comment]
Here's a project I did that's not just about WebSockets and Next.js, but the code should give you an idea of how to setup WebSockets with Next.js in any case, and as @gian412 mentions, you'll need to use a custom server setup:

https://github.com/dfjs/realtime-physical-spaces

You'll specifically want to look here for server-side implementation + references: 

https://github.com/dfjs/realtime-physical-spaces/blob/main/server.js#L6

I didn't use anything too heavy for integrating WebSockets on the back or front ends i.e. [this WebSockets implementation for Node.js](https://www.npmjs.com/package/ws) and [a simple React hook 'react-use-websocket'](https://www.npmjs.com/package/react-use-websocket).

Front-end / client-side code here: https://github.com/dfjs/realtime-physical-spaces/blob/main/pages/index.js

**Reply by jacobbogers** [reply]
I am thinking this might be possible with "middleware" (it needs to handle the HTTP `upgrade`
 request request) and I think api endpoints are only called if the request is parsed and considered received already.  I am not sure if nextjs (i need to look into this) will close the socket, by itself, or honors "CONNECTION: KEEP-ALIVE". Keep you guys updated

**Comment by anuraagbarde** [comment]
I don't seem to understand how WebSocket can work in the next.js/vercel infra. If I refer to the T3 stack ( https://create.t3.gg/ ), All the backend works in serverless functions, which obviously die off and hence can not keep a live connection for a long time. I may be wrong, please correct me.

**Reply by GabenGar** [reply]
Yes, websocket requires an actual persistent server, NextJS can act as a proxy at best in this scenario. A very expensive proxy which also renders html templates.
It's way easier to just have a separate endpoint to the ws server and interact to it with the client script only.

**Comment by brhx** [comment]
I've just created a package, [next-plugin-websocket](https://github.com/sam3d/next-plugin-websocket), that adds support for WebSocket handlers in Next.js API routes with no extra configuration:

```ts

import { appRouter } from "@/server/routers/_app";
import { NextApiHandler } from "next";
import { NextWebSocketHandler } from "next-plugin-websocket";

export const socket: NextWebSocketHandler = (client, req) => {
  client.on("message", (msg) => {
    client.send(msg);
  });
};

const handler: NextApiHandler = (req, res) => {
  res.status(426).end();
};

export default handler;
```

It's still very much in beta, and I've only managed to test it on Next.js 13 (with and without `appDir`) and standalone output mode, but if there are any extra features or anything that isn't working please let me know and I'll take a look!

I've been using it in production without issue for a little while now without any issues 😊

**Reply by anurag-roy** [reply]
Impeccable timing! I was several tabs deep and just found this posted 25 minutes ago. 🤣

Looks exactly like what I was looking for, I'll try it out and let you know of issues if any. Thanks!

**Reply by gustavostz** [reply]
@sam3d Good approach, but how is it working? Vercel does not support WebSockets as they are stateless and have a maximum execution duration, making it impossible to maintain a WebSocket connection.

They even have a post describing this: https://vercel.com/guides/do-vercel-serverless-functions-support-websocket-connections

**Reply by brhx** [reply]
@gustavostz it's sadly not compatible with serverless deployment environments in general, nor can it be. WebSockets are inherently stateful.

It works for `next dev`, `next start`, and the standalone output mode. I've found success deploying it on containerised services like AWS Fargate.

**Reply by GermanJablo** [reply]
hey @sam3d is this syntactic sugar to make a http that opens a ws connection as [shown here](https://stackoverflow.com/a/71415736/10476393)?
If so, I ask you the same question that I left for that user: Would each user be creating a ws server with a http request? Isn't there a way to create a global and permanent ws server at deploy time?

**Reply by GabenGar** [reply]
Yes, by running a dedicated ws server. Aka not serverless.

**Reply by GermanJablo** [reply]
Yes, I mean is there a way using the NextJS APIs routes*

**Reply by GabenGar** [reply]
API routes are HTTP/HTTPS, so no.

**Comment by isaiahtaylor** [comment]
FYI - I finally got streaming to work on serverless functions so I can stream OpenAI. See: https://twitter.com/isaiah_p_taylor/status/1638216571507331078

**Comment by JustFly1984** [comment]
AWS has serverless Websockets API Gateway. deploy almost the same as lambda.

**Comment by Kamleshpaul** [comment]
For anyone looking for a detailed guide on using WebSocket in Next.js with the app router and Socket.IO, I’ve written an article that might help: [How to Use WebSocket in Next.js App Router with Socket.IO](https://kamleshpaul.com/posts/how-to-use-websocket-in-nextjs-app-router-with-socketio).  

The article covers:  
- Setting up WebSocket with Next.js App Router.  
- Integrating Socket.IO for real-time communication.  
- Handling server-side and client-side connections effectively.  

I hope this can save you some time and effort. Feedback and suggestions are always welcome!

**Comment by dzulhelmynazri** [comment]
Check this out

[https://vercel.com/docs/functions/websockets](https://vercel.com/docs/functions/websockets)
